### ISSUE TITLE
Data Driven Testing

### ISSUE THREAD

**Issue by janmotl** [scikit-learn-contrib/categorical-encoding#117]
I was thinking about applying data driven testing (ddt).

Why ddt: 
1. `test_encoders` contains a lot of copy-pasted code
2. `test_estimators` already loops over all encoders
3. ddt, in comparison to a simple loop, does not stop on the first error (unless you tell it to do so) and reports results for each subtest (not just for the whole loop)
4. ddt makes it easy to quickly increase test coverage
5. ddt makes it easy to test that all encoders in the library fulfill some basic requirements

But of course, there are disadvantages:
1. It makes it slightly harder to understand and write tests
2. I have encountered behaviors (two, to be exact) that could be considered valid but break uniform behaviour of the library. I do not know what to do - adjust the tests or the encoders?

Example of the [code](https://github.com/scikit-learn-contrib/categorical-encoding/files/2349193/test_ddt.py.zip).

Questions:
1. Should I continue in the effort to port portable tests under ddt while preserving the encoder specific tests?
2. If yes, is it ok to make a branch instead of a pull request? I hope it is going to be easier (for me) to maintain synchronicity between the development and master branch.

**Comment by wdm0006** [comment]
I think this makes sense (the testing part), and think in this context a feature branch sounds fine. I'd caution though that we should build these tests alongside the existing test suite for now, and merge in regularly, rather than have a super-long-lived development branch.

**Comment by janmotl** [comment]
Pushed the solution into `tests` branch.

Changes:
1. runtime decreased from 90 seconds to 38 seconds on my laptop (mostly because of decreasing the dataset size from 1000 rows to 100 rows)
2. line test coverage of encoders increased from ~80% to ~94% (may vary based on the used definition)

Two tests fail. They are described in issues #120 and #121.

**Comment by janmotl** [comment]
By mistake, I have pushed into 'master' instead of 'tests' branch. I deeply apologize. I suggest to revert to commits 2394855f03ece9823a6182854e98e5f9d3a32e66 or b1993e09ce8b5aa21e5b6ffad19ae64f25d9311b. 

Current state: All tests now pass. At least on my computer.

Errors reported by TravisCI:

1. `Failure: ImportError (No module named unittest2)` Resolvable by adding dependency into TravisCI. Somewhere.
2. `ValueError: minlength must be positive` No idea (and I am using pandas 0.23.4 just like TravisCI).
3. `TypeError: __init__() got an unexpected keyword argument 'handle_unknown'` No idea.

CircleCI fails as well.

I am not able to resolve these issues.

**Comment by wdm0006** [comment]
CircleCI always fails, its just pushing docs. the failure is just that there are no tests, so thats ok. 

TravisCI installs the packages in the install_requires in setup.py, but thats not appropriate for unittest2, just add that directly into the install.sh script in ci_scripts directory.

on (2), not sure. 

On (3), not sure. 

Do you have tracebacks for either of them?

**Comment by janmotl** [comment]
The last issue in TravisCI was resolved by removing a * import. Now I get it why everyone says not to use * imports...

@wdm0006, will you take care of CircleCI?

**Comment by janmotl** [comment]
CircleCI is failing during `sudo apt-get update` with:
```
Err http://ppa.launchpad.net precise/main Sources
  404  Not Found

Err http://ppa.launchpad.net precise/main amd64 Packages
  404  Not Found

Err http://ppa.launchpad.net precise/main i386 Packages
  404  Not Found

Ign http://ppa.launchpad.net precise/main Translation-en_US

Ign http://ppa.launchpad.net precise/main Translation-en

Fetched 7,938 kB in 9s (852 kB/s)
W: GPG error: https://apt.dockerproject.org ubuntu-precise Release: The following signatures couldn't be verified because the public key is not available: NO_PUBKEY F76221572C52609D
W: Failed to fetch http://ppa.launchpad.net/pdoes/ppa/ubuntu/dists/precise/main/source/Sources  404  Not Found

W: Failed to fetch http://ppa.launchpad.net/pdoes/ppa/ubuntu/dists/precise/main/binary-amd64/Packages  404  Not Found

W: Failed to fetch http://ppa.launchpad.net/pdoes/ppa/ubuntu/dists/precise/main/binary-i386/Packages  404  Not Found

E: Some index files failed to download. They have been ignored, or old ones used instead.

sudo apt-get update returned exit code 100

Action failed: sudo apt-get update
```
Knowing nothing better, I would suggest to check `/etc/apt/sources.list` and `/etc/apt/sources.list.d/*.list` and remove sources that no longer exists. Unfortunately, I think I do not have the necessary privileges to do the change.

**Comment by wdm0006** [comment]
I'll take a look at this, thanks for the investigation work @janmotl
